routes/admin.js

```diff
@@ -0,0 +1,248 @@
+    const [approved, rejected, pending, created] = await Promise.all([
+        req.db.users.count({ where: { status: 'approved' } }),
+        req.db.users.count({ where: { status: 'rejected' } }),
+        req.db.users.count({ where: { status: 'manual_review' } }),
+        req.db.users.count({ where: { status: 'created' } }),
+    ]);
+    return {
+        approved, rejected, pending, created
+    };
+});
+
+addHandler('/get_signup', async req => {
+    const { where } = req.body;
+    if (!where) {
+        throw new Error('Missing where statement');
+    }
+    const user = await req.db.users.findOne({ where });
+    if (!user) {
+        throw new Error('Unknown user');
+    }
+    const actions = await req.db.actions.findAll({ where: {
+        [Sequelize.Op.or]: [
+            { ip: user.ip },
+            { user_id: user.id },
+        ]
+    } });
+    // TODO: geoip lookup should be stored per signup in database
+    //       so that it is searchable
+    const location = req.geoip.get(user.ip);
+    return { user, actions, location };
+});
+
+addHandler('/list_signups', async req => {
+    const { limit, order, offset, filters } = req.body;
+    const query = {
+        order: order || [['created_at', 'DESC']],
+        limit: limit ? Math.min(limit, 100) : 10,
+    };
+    if (offset) {
+        query.offset = offset;
+    }
+    if (Array.isArray(filters) && filters.length > 0) {
+        const { or, like, and, gte, lte } = Sequelize.Op;
+        const andList = [];
+        for (const filter of filters) { // eslint-disable-line
+            const { name, value } = filter;
+            switch (name) {
+                case 'text':
+                    andList.push({ [or]: [
+                        { email: { [like]: `%${value}%` } },
+                        { username: { [like]: `%${value}%` } },
+                        { phone_number: { [like]: `%${value}%` } },
+                        { fingerprint: { [like]: `%${value}%` } },
+                    ] });
+                    break;
+                case 'status':
+                    andList.push({ status: value });
+                    break;
+                case 'ip':
+                    andList.push({ ip: value });
+                    break;
+                case 'from':
+                    andList.push({ created_at: { [gte]: new Date(value) } });
+                    break;
+                case 'to':
+                    andList.push({ created_at: { [lte]: new Date(value) } });
+                    break;
+                default:
+                    throw new Error(`Unknown filter: ${name}`);
+            }
+        }
+        query.where = { [and]: andList };
+    }
+    const [total, users] = await Promise.all([
+        req.db.users.count({ where: query.where }),
+        req.db.users.findAll(query),
+    ]);
+    return { total, users, query };
+});
+
+addHandler('/approve_signups', async req => {
```

Comment:
when multiple signups are approved in a single request, the response generated by this handler does not make it easy to relate a certain success or failure with a given signup. it'd be more usable if the response returned a good/no-good result for each signup ID.

Reply:
It does, you get an array of success/error for each id you pass with the same indices as the id passed

Reply:
that works well enough :)